packages/preact/src/hooks/useSpeechToText.ts

```diff
@@ -0,0 +1,110 @@
+import { useCallback, useRef, useState } from "preact/hooks"
+
+interface SpeechToTextOptions {
+  language?: string
+  interimResults?: boolean
+}
+
+interface StartListeningOptions {
+  onResult: ({ value, confidence }: { value: string; confidence: number }) => void
+  onError?: (error: string) => void
+}
+
+export type SpeechToText = {
+  listening: boolean
+  startListening: (options: StartListeningOptions) => void
+  stopListening: () => void
+}
+
+const SpeechRecognition = window.SpeechRecognition || window.webkitSpeechRecognition
+export const speechToTextSupported = !!(SpeechRecognition && typeof SpeechRecognition === "function")
+
+function useSpeechToTextUnsupported(): SpeechToText {
+  return {
+    listening: false,
+    startListening: () => {},
+    stopListening: () => {}
+  }
+}
+
+function useSpeechToTextSupported({
+  language = "en-US",
+  interimResults = false
+}: SpeechToTextOptions = {}): SpeechToText {
+  const [listening, setListening] = useState(false)
+
+  const recognizerRef = useRef<SpeechRecognition | null>(null)
+
+  const startListening = useCallback(
+    ({ onResult, onError }: StartListeningOptions) => {
+      const recognizer = new SpeechRecognition()
+      recognizer.lang = language
+      recognizer.interimResults = interimResults
+
+      recognizer.onstart = () => setListening(true)
+
+      recognizer.onresult = event => {
+        const { transcript, confidence } = event.results?.[0]?.[0]
+        onResult({ value: transcript, confidence })
```

Comment:
how about

`onResult(transcript)`

and we can add confidence handling later if needed